src/collapse/mod.rs

```diff
@@ -1,2 +1,63 @@
 pub mod bpftrace;
 pub mod perf;
+
+use std::io;
+use std::path::Path;
+use std::fs::File;
+
+const READER_CAPACITY: usize = 128 * 1024;
+
+/// Type that knows how to collapse formats from various frontends (e.g. bpftrace and perf)
+/// into output that can be subsequently ingested by `inferno-flamegraph`.
+pub struct Collapser {
```

Comment:
I think this can probably be a trait instead:
```rust
trait Frontend {
    fn collapse<R: BufRead, W: Write>(r: R, w: W) -> io::Result<()>;
}
```
And then there'll be a free-standing function:
```rust
fn collapse_with<P, F>(infile: Option<P>, frontend: F) -> io::Result<()>
  where P: AsRef<Path>, F: Frontend {
    // ...
    frontend.handle(/* ... */)
    // ...
}
```
That way, each `main` just becomes something along the lines of:
```rust
collapser::collapse_with(infile.as_ref(), Perf::from(Opt::from_args()));
```
It'll require a little bit of re-organization in the perf module to have a `Perf` that implements `From<perf::Options>`, but that shouldn't be too bad.

What do you think?

Reply:
Yes. I like that. Wasn't sure how much you wanted me to mess with the existing perf code, but I like the idea of turning frontends into structs that all implement common behavior (i.e. a `Frontend` trait). Not sure `collapse_with` needs to be a standalone function (seems like it could be an autoimplemented method on the trait itself), but I'll play around with it.

Reply:
I like this architecture, although we may want to think of better names for `collapse` and `collapse_with`... It's a little non-obvious what the difference between the two is. I put that off just to get your feedback on this general structure. 

Also - I'm pretty sure the Windows issue is due to git changing to `\r\n` line endings for our data files on Windows. Let me know if you think otherwise, but since bpftrace only runs on windows, I think we should be ingesting and outputing files with unix-style line endings always, even on Windows. I think the perl script is always outputing `\n` line-endings, but, as I mentioned before, I don't really know perl.

Reply:
Also - Another unrelated thought as I was looking at the perf code. We're essentially interning (I think that's the right word) the `addr2line::Context`s because they're expensive to construct. I know this is generally a bad thing, but what do you think about making that HashMap a global variable via lazy_static (would probably have to be protected by a mutex, which is the downside). Benefit would be two-fold: 

* We could derive `Clone`, `Debug`, `Eq`, `PartialEq`, and `Hash` (I believe) for `Perf`, which are not strictly necessary but generally good to provide on public types if we can (who knows how people will want to use your library).
* If for some crazy reason people use multiple instances of `Perf` in their program, you'll save on constructing `addr2line::Context`s.

Reply:
> we may want to think of better names for `collapse` and `collapse_with`

Mmm, good point... How about `collapse` and `collapse_file`?

> since bpftrace only runs on windows

I'm confused? `bpftrace` is mostly a Linux tool as far as I'm aware? Is that what you meant?

> I think we should be ingesting and outputing files with unix-style line endings always, even on Windows.

Yes, I think I agree with this.

> We're essentially interning the `addr2line::Context`s because they're expensive to construct. [...] [W]hat do you think about making that HashMap a global variable via lazy_static.

Sadly this doesn't quite work, since you could have multiple files with the same module names but different `Context`s. If you share the map across multiple `Perf`s, that would mess up your inlining info. I'm not entirely sure that there's an advantage to deriving those types on `Perf`. It doesn't really make sense to do any of those operations on them I think, in part especially because they also need access to things like debug symbols!

Reply:
"since bpftrace only runs on windows" - Yah, I meant linux.